modules/geoipbackend/geoipbackend.cc

```diff
@@ -124,172 +125,240 @@ void GeoIPBackend::initialize() {
-          rr.content = content;
-          rr.weight = 100;
-        } 
-        rr.auth = 1;
-        rr.d_place = DNSResourceRecord::ANSWER;
-        rrs.push_back(rr);
-      }
-      std::swap(dom.records[qname], rrs);
+        else {
+          L << Logger::Info << "[GEOIPBACKEND][Parsing Zone File]: "<< ZoneFile << std::endl;
+          loadGeoFile(ZoneFile, tmp_domains);
+          
+       }
     }
+  }
+  else {
+    L << Logger::Info << "[GEOIPBACKEND][Parsing Zone File]: "<< getArg("zones-file") << std::endl;
+    std::string zones_file = std::string(getArg("zones-file"));
+    loadGeoFile(zones_file, tmp_domains);
+  }
 
-    for(YAML::const_iterator service = domain["services"].begin(); service != domain["services"].end(); service++) {
-      NetmaskTree<vector<string> > nmt;
+  s_domains.clear();
+  std::swap(s_domains, tmp_domains);
+}
 
-      // if it's an another map, we need to iterate it again, otherwise we just add two root entries.
-      if (service->second.IsMap()) {
-        for(YAML::const_iterator net = service->second.begin(); net != service->second.end(); net++) {
-          vector<string> value;
-          if (net->second.IsSequence()) {
-            value = net->second.as<vector<string> >();
-          } else {
-            value.push_back(net->second.as<string>());
-          }
-          if (net->first.as<string>() == "default") {
-            nmt.insert(Netmask("0.0.0.0/0")).second.assign(value.begin(),value.end());
-            nmt.insert(Netmask("::/0")).second.swap(value);
-          } else {
-            nmt.insert(Netmask(net->first.as<string>())).second.swap(value);
-          }
-        }
-      } else {
-        vector<string> value;
-        if (service->second.IsSequence()) {
-          value = service->second.as<vector<string> >();
-        } else {
-          value.push_back(service->second.as<string>());
+void GeoIPBackend::gatherZones(std::vector<std::string> &extraZones) {
+  extraZones.clear();
+  if (getArg("zones-dir").empty()) return; // nothing to do
+    struct stat st;
+    DIR *dir;
+    struct dirent *ent;
+
+    // stat
+    if (stat(getArg("zones-dir").c_str(), &st)) {
+       L << Logger::Error << getArg("zones-dir") << " does not exist!" << std::endl;
+       throw PDNSException(getArg("zones-dir") + " does not exist!");
+    }
+
+    // wonder if it's accessible directory
+    if (!S_ISDIR(st.st_mode)) {
+       L << Logger::Error << getArg("zones-dir") << " is not a directory" << std::endl;
+       throw PDNSException(getArg("zones-dir") + " is not a directory");
+    }
+
+    if (!(dir = opendir(getArg("zones-dir").c_str()))) {
+       L << Logger::Error << getArg("zones-dir") << " is not accessible" << std::endl;
+       throw PDNSException(getArg("zones-dir") + " is not accessible");
+    }
+
+    while((ent = readdir(dir)) != NULL) {
+      if (ent->d_name[0] == '.') continue; // skip any dots
+      if (boost::ends_with(ent->d_name, ".yml")) {
```

Comment:
these two could be combined as single if... `if (ent->d_name[0] != '.' && boost::ends_with(ent->d_name, ".yml") {`